Add support for generating LuaCATS definitions for the Lua API

kalimag opened the pull request:
Adds the ability to generate API definitions in the LuaCATS format to the Lua Console.

[LuaCATS](https://luals.github.io/wiki/annotations/) is primarily the annotation format of [Lua Language Server](https://luals.github.io/), which drives the most popular VS Code Lua extension and can be used in other editors with language server support.

~~LuaCATS is an extension/fork of the annotation format used by EmmyLua, but I'm not sure what the status of the latter is. These definitions might be compatible with EmmyLua, but the only spec of its format I could find is [sparse and partially in Chinese](https://emmylua.github.io/annotation.html).~~ EmmyLua extension now uses LuaCATS annotations, too.

The generated definitions could be further improved, e.g. enumerating valid values for `surfaceName` parameters, specifying the structure of certain `table` values, and callback function signatures. But that would require extending the API metadata and I wanted to get some feedback first.

Current output:
- [BizHawk Dev Builds](https://gist.github.com/kalimag/901d9e076e74668611f5872e9e48b87d)
- [BizHawk 2.11](https://gist.github.com/kalimag/a45b110239d75aa4a1ffa88c6a90cc8c)

Check if completed:
- [x] I, the committer, have read the [licensing terms for contributors](https://github.com/TASEmulators/BizHawk/blob/master/contributing.md#copyrights-and-licensing) (last updated 2022-07-15) and am compliant

will resolve #3860

kalimag committed:
Add support for generating LuaCATS definitions for the Lua API

kalimag commented:
Just for example, some issues LuaLS can detect in the bundled scripts using these definitions:

![image](https://github.com/TASEmulators/BizHawk/assets/14197345/63cb33e8-697f-4787-85cd-0b074e3d87fb)

![image](https://github.com/TASEmulators/BizHawk/assets/14197345/9d4fbd2d-a68d-4550-aa2a-7c91d8bc5ad1)

kalimag committed:
Add missing numeric types

kalimag committed:
Throw error when definition file is executed as a script

kalimag committed:
Add support for bytes-as-string parameters

vadosnaprimer committed:
Apply suggestions from code review

Co-authored-by: feos <vadosnaprimer@users.noreply.github.com>

vadosnaprimer committed:
wip

vadosnaprimer committed:
add examples and hawk info

I don't know regex, replacements probably can be done in a neater way

vadosnaprimer committed:
line endings

vadosnaprimer committed:
move editor/ide integration to its own menu

we gotta promote this!

YoshiRulz reviewed (changes requested):
After fixing malformed examples, works w/ SumnekoLua plugin in IDEA.

vadosnaprimer committed:
does "Braces XOR single-line" mean this?

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/Lua/LuaConsole.cs:
```diff
@@ -1184,6 +1184,18 @@ private void RegisterNotePadMenuItem_Click(object sender, EventArgs e)
+				? Path.GetDirectoryName(LuaImp.ScriptList.Filename)
+				: Config!.PathEntries.LuaAbsolutePath();
+
+			if (this.ShowFileSaveDialog(initDir, initFileName: "BizHawk.lua", fileExt: ".lua", filter: JustScriptsFSFilterSet) is string path)
```
[Per docs](https://luals.github.io/wiki/definition-files/#creating-definition-files)

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/Lua/LuaConsole.cs:
```diff
@@ -1184,6 +1184,18 @@ private void RegisterNotePadMenuItem_Click(object sender, EventArgs e)
+		{
+			string initDir = !string.IsNullOrWhiteSpace(LuaImp.ScriptList.Filename)
+				? Path.GetDirectoryName(LuaImp.ScriptList.Filename)
+				: Config!.PathEntries.LuaAbsolutePath();
```
Moving the menu up for discoverability sounds good, but I don't see why users should have to generate these for themselves.

vadosnaprimer commented on src/BizHawk.Client.Common/lua/Documentation/LuaCatsGenerator.cs:
```diff
@@ -0,0 +1,187 @@
+
+		sb.AppendLine(Preamble);
+
+		foreach (var libraryGroup in docs.GroupBy(func => func.Library).OrderBy(group => group.Key))
```
How did you install it exactly then?

vadosnaprimer committed:
braces

vadosnaprimer committed:
name the definitions file according to Lua Language Server spec

vadosnaprimer committed:
proper Program Files folder search

kalimag committed:
Refactor example formatting

kalimag committed:
Preamble, line-ending stuff

kalimag committed:
Make `LuaCanvas` local

kalimag committed:
Refactor `GetLuaType(ParameterInfo)`

kalimag committed:
Enumerate valid drawingSurface values

kalimag committed:
It's free braces

kalimag committed:
Make class static

kalimag committed:
Make class static Part II

YoshiRulz commented on src/BizHawk.Client.EmuHawk/tools/Lua/LuaAutocompleteInstaller.cs:
```diff
@@ -46,17 +46,19 @@ public void InstallBizLua(TextEditors editor, LuaDocumentation docs)
 		{
 			// The most likely location of the app, eventually we should consider looking through the registry or installed apps as a more robust way to detect it;
-			string exePath = @"C:\Program Files (x86)\Notepad++\notepad++.exe";
+			string exePath = Path.Combine(ProgramFilesFolder, @"Notepad++\notepad++.exe");
```
Will this be finished in this PR? Because there's no point touching it if it's going to stay broken, just set `Visible = false` on the menu items and it can continue gathering dust.

vadosnaprimer committed:
split into individual files?

revert if we don't want this anymore, but with LLS's ability to take in entire folders via config this feels nicer IMO

and yeah I wouldn't mind adding this output to releases under something like `_luacats_definitions` folder in `Assets/Lua`

kalimag committed:
Remove return statement from definition files

These are globals rather than `require`'d imports

kalimag committed:
Include default parameter values

kalimag committed:
Fix code style

YoshiRulz commented on src/BizHawk.Client.Common/lua/Documentation/LuaCatsGenerator.cs:
```diff
@@ -0,0 +1,234 @@
+		throw new NotSupportedException($"Unknown type {type.FullName} used in API. Generator must be updated to handle this.");
+	}
+
+	private static bool IsNullable(Type type) => type.IsGenericType && type.GetGenericTypeDefinition() == typeof(Nullable<>);
```
All reference types in the Lua API are nullable because Lua is not a .NET language and cannot "see" the NRT metadata.

kalimag replied:
That's the point, though? `gui.addmessage(nil)` throws an error, so the diagnostics should warn about it. `event.onexit(function() end, nil)` doesn't throw an error, so diagnostics shouldn't warn.

That both are a nullable reference type on the technical level is irrelevant. As far as Lua is concerned, anything goes and you can pass a dozen table arguments to `memory.writebyte` if you want, but the diagnostics should help you catch that before it causes a runtime error.

kalimag replied:
Reference types are now annotated as nullable, unless they are compiled with `#nullable enable` (and not marked as nullable in C#).

Still need to do the legwork of adding `#nullable` to existing Lua libs.

YoshiRulz replied:
Wait they are available with reflection? I was under the impression that you needed `NullabilityInfoContext` from .NET Core.
I'll see about removing the need for `dynamic`.

kalimag replied:
Sure, it just gets rather complicated to correctly interpret the attribute data without `NullabilityInfoContext` once you get into generics and nested arrays and whatever. But for this purpose it's simple enough.

YoshiRulz replied:
Not sure what you mean.
Pushed https://github.com/TASEmulators/BizHawk/commit/227bb47d7f1ef854c2ba99d0cac2c7693c845fd2. It's mostly the same as what you had except it returns `null` in the fallback (for `#nullable disable`), so you may need `?? true` at the callsites.
I'll be adding it into the TASVideos wiki generator too (edit: #4688).
Calling this resolved pending a rebase, and the work of adopting NRTs in the Lua libraries can come later.

kalimag committed:
Rename color aliases

kalimag committed:
Fix name collision with `userdata`

kalimag committed:
Mark zero-indexed return values

YoshiRulz commented:
Please rebase.
Do you mind if I add `[LuaZeroIndexed]` myself? It could be used in the existing TASVideos wiki markup, and those functions' current documentation make no mention of it. edit: fdec9a3b844d75f2174038efd4a3d419f1b923f5

kalimag committed:
Handle zero-indexed parameters

kalimag committed:
Support Nullable Reference Type annotations

kalimag committed:
Revert "wip" - unrelated submodule change

This reverts commit c9fbcdcfc7c436bb0daa5f9053fcb6549519e475.

kalimag committed:
Switch to NRT extensions

YoshiRulz reviewed (approved):
Works great, code LGTM. If I don't return to finish #4688 soon enough then by all means merge both PRs. I think nullable `joypad.get` would be the most annoying false positive.

vadosnaprimer commented:
How soon is "soon enough"? I'd like to release within a week.

SuuperW commented:
When I hover over a gui.draw method in VS Code to see the documentation, `color` is replaced with `string|integer|dotnetcolor` which I don't think is good. If there's a way for the definition to hint to IDEs that expansion isn't desired I think that'd be good, but as far as I can tell there doesn't seem to be a way to do that.

I can change the settings in the VS Code extension to disable alias expansion at least.

vadosnaprimer commented:
Maybe making a derived class instead of alias would do what you want, but I personally don't see a problem in it showing the expanded thing.

kalimag commented:
>I can change the settings in the VS Code extension to disable alias expansion at least.

Yeah I don't think the annotations have any control over that. That's up to the IDE and individual users' settings. EmmyLua doesn't do this by default, FWIW.

> Maybe making a derived class instead of alias would do what you want

No, that wouldn't work because a class can't inherit a union of types.

kalimag committed:
Adjust annotation whitespace and comment

YoshiRulz commented:
Can you put `dotnetcolor` first in the union?

kalimag committed:
Sort color union

kalimag committed:
Annotate more specific `createcanvas` return type
